### PR TITLE
Completely rework GLB loader and add from_reader method

### PR DESCRIPTION
**Pull request by fst3a**
This PR supersedes #123 but is somewhat more aggressive.  It completely reworks GLB loader to make code more clear to understand and also implements a `from_reader` method in addition to existing `from_slice`, which takes responsibility of loading GLB from any `io::Read` stream for extra user convenience.

### CODE REVIEW

File: src/glb.rs [hunk 1]
```diff
@@ -6,47 +6,118 @@
...
 // except according to those terms.
+use byteorder::{LE, ReadBytesExt};
 
 use Error;
 
+use std::{fmt, io};
+
+/// Represents a Glb loader error.
+#[derive(Debug)]
+pub enum GlbError {
```
Comment (alteous) [inline src/glb.rs]: I have an aversion toward repeated prefixes and suffixes. `Error` is just fine, as the crate will use `glb::Error` and the user `gltf::glb::Error`.

File: src/glb.rs [hunk 2]
```diff
@@ -6,47 +6,118 @@
...
+    /// Io error occured.
+    IoError(::std::io::Error),
```
Comment (alteous) [inline src/glb.rs]: Same here. We know it's an error, so `Io` is enough.

Comment (alteous) [reply src/glb.rs]: I like the breakdown of the individual problems.

File: src/glb.rs [hunk 3]
```diff
@@ -6,47 +6,118 @@
...
     pub version: u32,
-
     /// Must match the length of the parent .glb file.
     pub length: u32,
 }
 
-/// The header of the JSON or BIN section of a .glb file.
+/// GLB chunk type.
+#[derive(Copy, Clone, Debug)]
+pub enum ChunkType {
```
Comment (alteous) [inline src/glb.rs]: :+1:

File: src/glb.rs [hunk 4]
```diff
@@ -6,47 +6,118 @@
...
+    Json,
+    /// BIN\0 chunk.
+    Bin,
+}
+
+/// Chunk header with no data read yet.
 #[derive(Copy, Clone, Debug)]
-#[repr(C)]
-struct ChunkInfo {
+struct ChunkHeader {
```
Comment (alteous) [inline src/glb.rs]: This is better, thanks.

File: src/glb.rs [hunk 5]
```diff
@@ -6,47 +6,118 @@
...
+                length: reader.read_u32::<LE>().map_err(IoError)?,
+            })
+        } else {
+            Err(GlbError::Magic(magic))
+        }
+    }
 
-    /// Either `b"JSON"` or `b"BIN\0"`.
-    kind: [u8; 4],
+    fn size_of() -> usize { 12 }
```
Comment (alteous) [inline src/glb.rs]: :+1:

File: src/glb.rs [hunk 6]
```diff
@@ -55,71 +126,134 @@ impl<'a> Glb<'a> {
...
-            }
-            if (*x).length as usize != data.len() {
-                return err("length does not match length of data");
-            }
-            if (*x).version != 2 {
-                return err("Not GLB version 2.0");
+    /// Does the loading job for you.  Provided buf will be cleared before new
+    /// data will be written.  When error happens, if only header was read, buf
+    /// will not be mutated, otherwise, buf will be empty.
+    pub fn from_reader<R: io::Read>(mut reader: R,
```
Comment (alteous) [inline src/glb.rs]: Good idea. :+1:

File: src/glb.rs [hunk 7]
```diff
@@ -55,71 +126,134 @@ impl<'a> Glb<'a> {
...
+                // SAFETY: We are doing unsafe operation on a user-supplied
+                // container!  Make sure not to expose user to uninitialized
+                // data if an error happens during reading.
+                //
+                // It is guaranteed by reserve's implementation that the reserve
+                // call will make buf's capacity _at least_ buf_len.
+                //
+                // We do not read contents of the Vec unless it is fully
+                // initialized.
+                unsafe { buf.set_len(glb_len as usize) };
```
Comment (alteous) [inline src/glb.rs]: The safety note is a bit overkill but I appreciate the sentiment.

File: src/glb.rs [hunk 8]
```diff
@@ -55,71 +126,134 @@ impl<'a> Glb<'a> {
...
+            .and_then(|json_h| if json_h.length as usize <= data.len() {
+                Ok(json_h)
+            } else {
+                Err(GlbError::ChunkLength {
+                    ty: json_h.ty,
+                    length: json_h.length,
+                    length_read: data.len(),
+                })
+            })
+            // PANIC: We have verified that json_h.length is no greater than
```
Comment (alteous) [inline src/glb.rs]: I'm a bit worried to be reading `PANIC` in big letters in the source code. Does this actually `panic!`?

File: src/glb.rs [hunk 9]
```diff
@@ -55,71 +126,134 @@ impl<'a> Glb<'a> {
...
+             GlbError::Version(_) => "unsupported version",
+             GlbError::Magic(_) => "not glTF magic",
+             GlbError::Length { .. } => "could not completely read the object",
+             GlbError::ChunkLength { ty, .. } => match ty {
+                 ChunkType::Json => "JSON chunk length exceeds that of slice",
+                 ChunkType::Bin => "BIN\\0 chunk length exceeds that of slice",
+             },
+             GlbError::ChunkType(ty) => match ty {
+                 ChunkType::Json => "was not expecting JSON chunk",
+                 ChunkType::Bin => "was not expecting BIN\\0 chunk",
```
Comment (alteous) [inline src/glb.rs]: The `\\0` part is irrelevant to the user-facing error but it's no big deal.

File: src/lib.rs [hunk 10]
```diff
@@ -125,10 +126,8 @@ pub use self::texture::Texture;
 pub enum Error {
     /// JSON deserialization error.
     Deserialize(json::Error),
-
     /// GLB parsing error.
-    Glb(String),
-
+    Glb(self::glb::GlbError),
```
Comment (alteous) [inline src/lib.rs]: Would prefer `Glb(glb::Error)` as discussed at the top.

File: src/glb.rs (hunk 8 above)
Comment (fst3a) [reply src/glb.rs]: No, it explains why it doesn't.

File: src/glb.rs (hunk 9 above)
Comment (fst3a) [reply src/glb.rs]: I understand, but this is made to not to confuse the user, so that user could easily look up BIN\0 chunk type in e.g. gltf specs.

**Comment by fst3a** [comment]
I will try to come up with a fix for AccessorIter, but AccessorIter is a part of separate crate (gltf-utils) anyway.

File: src/glb.rs (hunk 8 above)
Comment (alteous) [reply src/glb.rs]:
```rust
/// FIRE MISSILES: This function does not fire any missiles.
fn foo() {}
```

**Comment by fst3a** [comment]
This satisfies most requested changes except sound AccessorIter (part of gltf-utils).

**Comment by alteous** [comment]
Did you see my PR?

**Comment by alteous** [comment]
([This one](https://github.com/fst3a/gltf/pull/1))

**Comment by fst3a** [comment]
Whoops, no, I didn't see it.

**Comment by fst3a** [comment]
Don't really know how to test AccessorIter so implementation might be incorrect.

**Comment by fst3a** [comment]
Would be nice for somebody else to double-check the implementation against spec and/or express their concerns.  Don't really know how to test this for edge cases.

**Comment by alteous** [comment]
Will double-check with `three` and `gltf-viewer`.

**Comment by fst3a** [comment]
I'm certainly sure we never use signed integers in gltf-utils, even though they appear in the spec, though.  Other than that, we are apparently missing some matrix types?

type | Number of components
-- | --
"SCALAR" | 1
"VEC2" | 2
"VEC3" | 3
"VEC4" | 4
"MAT2" | 4
"MAT3" | 9
"MAT4" | 16

Should AccessorItem be implemented for those as well?

**Comment by fst3a** [comment]
It doesn't appear that we use any signed types anyway, even though they appear in spec.  As for spec it only requires us to have i8 and i16 but not i32 so I didn't implement it.

As for matrices, we're already covered by array implementations.  [[T; 4]; 4] would work because [T; 4] has AccessorItem implemented for it so no need to explicitly implement it for the nested type.

**Comment by alteous** [comment]
Awesome work! Thank you. :smile:

**Comment by fst3a** [comment]
This fixes #120 and #121 .
